clients/src/test/java/org/apache/kafka/common/telemetry/internals/SinglePointMetricTest.java

```diff
@@ -0,0 +1,162 @@
+import java.util.Collections;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class SinglePointMetricTest {
+
+    private MetricKey metricKey;
+    private Instant now;
+
+    /*
+     Test compares the metric representation from returned builder to ensure that the metric is
+     constructed correctly.
+
+     For example: Gauge metric with name "name" and double value 1.0 at certain time is represented as:
+
+       name: "name"
+          gauge {
+            data_points {
+              time_unix_nano: 1698063981021420000
+              as_double: 1.0
+            }
+          }
+     */
+
+    @BeforeEach
+    public void setUp() {
+        metricKey = new MetricKey("name", Collections.emptyMap());
+        now = Instant.now();
+    }
+
+    @Test
+    public void testGaugeWithNumberValue() {
+        SinglePointMetric gaugeNumber = SinglePointMetric.gauge(metricKey, Long.valueOf(1), now);
+        assertEquals(getIntGaugeMetric("name", 1, now),
+            gaugeNumber.builder().build());
+    }
+
+    @Test
+    public void testGaugeWithDoubleValue() {
+        SinglePointMetric gaugeNumber = SinglePointMetric.gauge(metricKey, 1.0, now);
+        assertEquals(getDoubleGaugeMetric("name", 1.0, now),
+            gaugeNumber.builder().build());
+    }
+
+    @Test
+    public void testGaugeWithMetricTags() {
+        MetricKey metricKey = new MetricKey("name", Collections.singletonMap("tag", "value"));
+        SinglePointMetric gaugeNumber = SinglePointMetric.gauge(metricKey, 1.0, now);
+        assertEquals(getDoubleGaugeMetric("name", 1.0, now,
+                Collections.singletonMap("tag", "value")), gaugeNumber.builder().build());
+    }
+
+    @Test
+    public void testSum() {
+        SinglePointMetric sum = SinglePointMetric.sum(metricKey, 1.0, false, now);
+        assertEquals(getSumMetric("name", 1.0, false, now), sum.builder().build());
+    }
+
+    @Test
+    public void testSumWithStartTimeAndTags() {
+        MetricKey metricKey = new MetricKey("name", Collections.singletonMap("tag", "value"));
+        SinglePointMetric sum = SinglePointMetric.sum(metricKey, 1.0, true, now, now);
+        assertEquals(getSumMetric("name", 1.0, true, now, now,
+            Collections.singletonMap("tag", "value"), false), sum.builder().build());
+    }
+
+    @Test
+    public void testDeltaSum() {
+        SinglePointMetric sum = SinglePointMetric.deltaSum(metricKey, 1.0, true, now, now);
+        assertEquals(getSumMetric("name", 1.0, true, now, now, Collections.emptyMap(),
+            true), sum.builder().build());
+    }
+
+    private Metric getDoubleGaugeMetric(String name, Double value, Instant time) {
+        return getDoubleGaugeMetric(name, value, time, Collections.emptyMap());
+    }
+
+    private Metric getDoubleGaugeMetric(String name, Double value, Instant time, Map<String, String> tags) {
+        NumberDataPoint.Builder point = NumberDataPoint.newBuilder()
```

Comment:
Seems like this code is the same as the actually production code? Is there no simpler way to create an expected result?

Or should we change the test condition `assertEquals` to check if the fields of the computed `SinglePointMetric`  are set correctly?

Atm, the test is like (hope this illustrates what I mean
```
// production code
int add(int a, int b) {
 return a + b;
}

// test
void shouldAddNumbers() {
  int sum = add(5, 6);
  assertEquals(addNumbers(5,6), sum); // <- this should be assert(11, sum); instead of duplicating production code inside `addNumbers` ?
}

int addNumbers(int a, int b) {
  return a + b;
}
```

Reply:
Yes, you are right. I was thinking similar while writing but should have used the assertions directly. I have changed the test code, done.